### PR TITLE
Adds a new `dlq_custom_codes` option to customize DLQ codes

### LINKED ISSUE #697: Make 403 response configurable for going into DLQ
When a 403 response is encountered, this is retried forever. This blocks the pipeline from continuing to process (https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/140). This is good in the event of a major misconfiguration since the entire pipeline becomes blocked. This can be bad in the case where just a couple messages are routed to indices that might be closed out due to automatic curation. The behavior of a 403 messages should be configurable, so it could go into the DLQ.

```
[2017-10-24T15:31:30,232][INFO ][logstash.outputs.elasticsearch] retrying failed action with response code: 403 ({"type"=>"index_closed_exception", "reason"=>"closed", "index_uuid"=>"Lnvl85mrRtSOGDzZnb86gw", "index"=>"test_dlq"})
```

Discussed with @jordansissel, and agree this should not be the default behavior - since putting in the DLQ only extends the amount of time before the pipeline is stalled (due to a full DLQ). However, it should be a choice an end user can configure for desired behavior.

### PR DESCRIPTION
**Pull request by andsel**
## Release notes
Adds `dlq_custom_codes` option to customize the list of HTTP codes that are considered eligible for DLQ.

## What does this PR do?
Introduced a new configuration option, named `dlq_custom_codes` which accepts a list of HTTP error codes.
The list is to be considered as an addition to the already existing codes that the plugin uses (DLQ_CODES, DOC_SUCCESS_CODES, DOC_CONFLICT_CODE defined in module `Common`). If the list has an intersection with those error codes then it's reported as a configuration error.

## Why is it important/What is the impact to the user?
This is a feature requested in #697, in some cases some error codes returned by Elasticsearch should be handled as DLQ messaged and not retried for ever. For example writing to a read only index report a 403, the index status would never change automatically so it's best to move these events to DLQ for a later reprocessing.
Under normal conditions the user shouldn't customize this list, but it's an helper in some circumstances.

## Checklist

- [x] My code follows the style guidelines of this project
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [x] test a version of the plugin with a real ES

## How to test this PR locally
- launch an ES instance, for example `docker-compose -f es_and_kb.yml up` 
[es_and_kb.txt](https://github.com/logstash-plugins/logstash-output-elasticsearch/files/8081186/es_and_kb.txt)
- from http://localhost:5601 Kibana UI (credential: elastic, changeme) create and index and put into read only:
```
PUT /test_index
PUT test_index/_settings 
{ 
  "index.blocks.read_only" : true 
}
```
- configure a pipeline to write into:
```
input {
	stdin {}
}

output {
	elasticsearch {
		index => "test_index"
		hosts => "http://localhost:9200"
		user => "elastic"
		password => "changeme"
		dlq_custom_codes => [403]
	}
}	
```
- enable DLQ in `config/logstash.yml` : 
```
dead_letter_queue.enable: true
```
- start the pipeline and generate an input and stop
- check that the string you typed is present in DLQ 
```
strings data/dead_letter_queue/main/1.log
```

## Related issues

- Fixes #697
- Related https://github.com/elastic/logstash/issues/10493

## Use cases

HTTP error codes used to decide if an event index reply has to be moved into the DLQ are 400 and 404. These codes are hard coded, this PR aims the user to customize this list extending other than 400 and 404.

### CODE REVIEW

File: docs/index.asciidoc
Comment (yaauie) [inline docs/index.asciidoc:530]: While the per-action response codes _mirror_ http status codes, we only handle retries and DLQ routing for failed actions from bulk requests whose HTTP status code is 2xx. This has caused a fair bit of confusion in the past, so I believe we need to be careful to call these out as action response codes and avoid any explicit association with HTTP.

File: lib/logstash/plugin_mixins/elasticsearch/common.rb
Comment (yaauie) [inline lib/logstash/plugin_mixins/elasticsearch/common.rb:271]: I would prefer using a `Set` (loaded once at plugin registration) that contained the union of these two, loaded into an instance variable, instead of adding a per-event overhead.

File: lib/logstash/outputs/elasticsearch.rb
Comment (kaisecheng) [inline lib/logstash/outputs/elasticsearch.rb:553]: besides success, conflict and dlq codes, there is a `failure_type_logging_whitelist`. Do you think it should be checked here?

File: lib/logstash/outputs/elasticsearch.rb
Comment (kaisecheng) [inline lib/logstash/outputs/elasticsearch.rb:549]: suggest a rephase for you to consider "#{intersection} are success codes which cannot be redefined in dlq_custom_codes"

File: docs/index.asciidoc
Comment (andsel) [reply docs/index.asciidoc:530]: I think is sufficient to rephrase it as:
```
List response or single action error codes that...
```
WDYT?

File: lib/logstash/outputs/elasticsearch.rb
Comment (kaisecheng) [inline lib/logstash/outputs/elasticsearch.rb:261]: about the naming, I understand there are fixed codes 400 and 404, so name the changeable part as custom codes. As a user, I think `dlq_codes` is more approachable and easy to understand.

File: docs/index.asciidoc
Comment (yaauie) [reply docs/index.asciidoc:530]: This specific line here, but we also have references to HTTP codes throughout.

```suggestion
List single-action error codes from Elasticsearch's Bulk API that are considered valid to move the events into the dead letter queue.
```

File: docs/index.asciidoc
Comment (yaauie) [inline docs/index.asciidoc:532]: It looks like we already handle `200`/`201` as explicit success, `400`/`404` as DLQ, and `409` as conflict, and are declaring it a configuration error to re-use any of these.

If a user can't use this to intercept `409` conflicts, or 404's attempting to update records that don't exist, what error codes do we expect? Only authorization-related ones like `403`? Others? Do we have real-world use-cases driving this feature?

File: lib/logstash/outputs/elasticsearch.rb
Comment (yaauie) [reply lib/logstash/outputs/elasticsearch.rb:261]: I favor `dlq_custom_codes`, because I would assume `dlq_codes` to require passing a complete list to _override_ the list of codes we naturally route to DLQ, whereas the `custom` bit implies that it is in _addition_ to the natural routing.

File: lib/logstash/outputs/elasticsearch.rb
Comment (yaauie) [inline lib/logstash/outputs/elasticsearch.rb:301]: if dlq is not enabled for a pipeline, and a user explicitly tells us how to route things to the DLQ, I would consider that a configuration error too.

File: docs/index.asciidoc
Comment (andsel) [reply docs/index.asciidoc:532]: This PR aim to solve request #697 which explicitly state that `403` error code could be eventually added as a DLQ trigger code.
While intersection with existing DLQ codes (`400` and `404`) could be seen as not necessarily a conflict, because both trigger the same feature, using error codes already configured as success (`200`/`201`) document conflict error (`409`) generates confusion in the precedence order when the [code process responses](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/1cab1ad2a41b28ac569cb77809b48ed6b66d2b39/lib/logstash/plugin_mixins/elasticsearch/common.rb#L264-L280).

Suppose that also `409` is customized as a `dlq_custom_code` what should take precedence? Processing it as a failed action or as DLQ event?

To keep the things simple I imagined to keep out of the set the codes that already has a meaning in the plugin.

File: lib/logstash/outputs/elasticsearch.rb
Comment (andsel) [reply lib/logstash/outputs/elasticsearch.rb:261]: The intention was to inform the user that's not the full overriding list but an addition list to the default behavior.

File: lib/logstash/outputs/elasticsearch.rb
Comment (andsel) [reply lib/logstash/outputs/elasticsearch.rb:553]: `failure_type_logging_whitelist` is used to log or not errors during the execution of the queries, while the method `#check_dlq_custom_codes` is used during registration to report configuration error. I think are different aspect, one concern about a configuration problem that should be logged or raised independently of `failure_type_logging_whitelist` which is a setting used to limit the number of logs.

File: lib/logstash/outputs/elasticsearch.rb
Comment (andsel) [reply lib/logstash/outputs/elasticsearch.rb]: Fixed by 5a81f0a328dbdcaa683cf7fd059677f3791a0dbe

File: lib/logstash/outputs/elasticsearch.rb
Comment (andsel) [reply lib/logstash/outputs/elasticsearch.rb:301]: nice point!

Comment (andsel) [reply lib/logstash/outputs/elasticsearch.rb:301]: Fixed by 8e98c89a7d618698e7d2871001d39bec44857fa5

File: lib/logstash/plugin_mixins/elasticsearch/common.rb
Comment (andsel) [reply lib/logstash/plugin_mixins/elasticsearch/common.rb]: Fixed by ec57f24c27fdb595b9c78e13a2c67ce754fd7926

File: lib/logstash/plugin_mixins/elasticsearch/common.rb
Comment (kaisecheng) [inline lib/logstash/plugin_mixins/elasticsearch/common.rb:271]: It is causing red CI in integration test version 6.x

**Review by kaisecheng — COMMENTED** [review]
Once the CI is fixed, I can approve.

File: lib/logstash/plugin_mixins/elasticsearch/common.rb
Comment (andsel) [reply lib/logstash/plugin_mixins/elasticsearch/common.rb:271]: the error is:
```
undefined method `include?' for nil:NilClass\nDid you mean?  include_class", :exception=>NoMethodError
```
So `@dlq_codes` is `nil` fo r `6.x`. I think it's related to the way `dlq_enabled?` works on LS `6.x`

Comment (andsel) [reply lib/logstash/plugin_mixins/elasticsearch/common.rb:271]: My guess was that the way `dlq_enabled?` operates on classes to determine if DLQ is active involved some fields that wasn't present on `6.8` but [ExecutionContextExt.dlqWriter](https://github.com/elastic/logstash/blob/50b43ebbaf35a042b9d1848118dcffde264dfbc9/logstash-core/src/main/java/org/logstash/execution/ExecutionContextExt.java#L39) exists in `6.8`

File: docs/index.asciidoc
Comment (roaksoax) [reply docs/index.asciidoc:532]: @yaauie seems that this needs your feedback?

Comment (yaauie) [reply docs/index.asciidoc:532]: Yes. I can unblock this PR.

We can consider a later feature to enable this explicit custom-code routing to also intercept `400`/`404`/`409` that are already routed by the implementation, and if and when we do so we have a path to do it in a non-breaking way.

**Review by yaauie — APPROVED** [review]
LGTM 👍🏼

**Comment by yaauie** [comment]
I've merged in upstream and resolved conflicting version bump in changelog/gemspec, and will merge/release on successful CI.

**Pull request merged**
